numpy/core/tests/test_regression.py

```diff
@@ -2214,6 +2214,20 @@ def test_reshape_size_overflow(self):
             new_shape = (2, 7, 7, 43826197)
         assert_raises(ValueError, a.reshape, new_shape)
 
+    def test_invalid_structured_dtypes(self):
+        # gh-2865
+        # the following invalid dtypes were causing segfaults
+        assert_raises(ValueError, np.dtype, ('O', [('name', 'i8')]))
+        assert_raises(ValueError, np.dtype, ('i8', [('name', 'O')]))
+        # these were causing SystemErrors
+        assert_raises(ValueError, np.dtype,
+                      ('i', {'name': ('i', 0, 'title', 'oops')}))
+        assert_raises(ValueError, np.dtype,
+                      ('i', {'name': ('i', 'wrongtype', 'title')}))
+        # this one is allowed as a special case though
+        a = np.ones(1, dtype=('O', [('name', 'O')]))
```

Comment:
Is it important that this example works? If not, I would be tempted to altogether disallow the `(base_dtype, new_dtype)` form of specification if either `base_dtype` or `new_dtype` have objects (tested with `PyDataType_REFCHK`). 

Note that similar operations are already disallowed:
```
>>> np.zeros(3, dtype='O').view(dtype([('name', 'O')]))
TypeError: Cannot change data-type for object array.
```

That wouldn't allow your special case though.